File: internal/pipe/pipe.go
package pipe

import (
	"bufio"
	"encoding/json"
	"os"
	"syscall"
	"time"
)

type Message struct {
	Type     string `json:"type"`
	JobID    string `json:"jobId,omitempty"`
	TaskID   string `json:"taskId,omitempty"`
	Message  string `json:"message,omitempty"`
	Question string `json:"question,omitempty"`
	Reason   string `json:"reason,omitempty"`
	Escalate bool   `json:"escalate,omitempty"`
	Artifacts []string `json:"artifacts,omitempty"`
}

// Create creates a named pipe at the given path.
func Create(path string) error {
	os.Remove(path) // clean up old pipe
	return syscall.Mkfifo(path, 0666)
}

// Listen opens the pipe and calls handler for each message.
// Re-opens on EOF to keep listening.
func Listen(path string, handler func(Message)) {
	for {
		f, err := os.Open(path)
		if err != nil {
			time.Sleep(100 * time.Millisecond)
			continue
		}

		scanner := bufio.NewScanner(f)
		for scanner.Scan() {
			var msg Message
			if json.Unmarshal(scanner.Bytes(), &msg) == nil {
				handler(msg)
			}
		}
		f.Close()
	}
}

// Send writes a message to the named pipe.
func Send(path string, msg Message) error {
	f, err := os.OpenFile(path, os.O_WRONLY, 0)
	if err != nil {
		return err
	}
	defer f.Close()

	data, err := json.Marshal(msg)
	if err != nil {
		return err
	}
	_, err = f.Write(append(data, '\n'))
	return err
}

File: pipe.ts
import * as fs from "fs";
import * as readline from "readline";
import * as child_process from "child_process";

export interface PipeMessage {
  type: string;         // "done" | "fail" | "ask" | "answer" | "new_task"
  jobId?: string;
  taskId?: string;
  message?: string;
  question?: string;
  reason?: string;
  escalate?: boolean;
  artifacts?: string[];
}

/** Create a named FIFO pipe at path. Removes stale pipe first. */
export function createPipe(pipePath: string): void {
  try { fs.unlinkSync(pipePath); } catch { /* ok */ }
  child_process.execFileSync("mkfifo", [pipePath]);
}

/**
 * Continuously listen on a named FIFO pipe and call handler for each message.
 * Re-opens on EOF (writer disconnected) — same behaviour as Go implementation.
 */
export async function listenPipe(
  pipePath: string,
  handler: (msg: PipeMessage) => void
): Promise<void> {
  while (true) {
    await new Promise<void>((resolve) => {
      let stream: fs.ReadStream;
      try {
        stream = fs.createReadStream(pipePath);
      } catch {
        setTimeout(resolve, 200);
        return;
      }

      const rl = readline.createInterface({ input: stream, crlfDelay: Infinity });

      rl.on("line", (line) => {
        const trimmed = line.trim();
        if (!trimmed) return;
        try {
          const msg = JSON.parse(trimmed) as PipeMessage;
          handler(msg);
        } catch { /* skip malformed */ }
      });

      rl.on("close", resolve);
      stream.on("error", () => resolve());
    });

    // Brief pause before reopening (allows OS to clean up)
    await sleep(100);
  }
}

/** Send a message to the foreman via named pipe. */
export function sendPipe(pipePath: string, msg: PipeMessage): void {
  const data = JSON.stringify(msg) + "\n";
  const fd = fs.openSync(pipePath, "w");
  try {
    fs.writeSync(fd, data);
  } finally {
    fs.closeSync(fd);
  }
}

function sleep(ms: number): Promise<void> {
  return new Promise((r) => setTimeout(r, ms));
}
